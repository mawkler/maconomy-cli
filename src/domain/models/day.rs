use std::{borrow::Borrow, fmt::Display, str::FromStr};

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) enum Day {
    Monday,
    Tuesday,
    Wednesday,
    Thursday,
    Friday,
    Saturday,
    Sunday,
}

impl FromStr for Day {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let day = match s.to_lowercase().borrow() {
            "monday" => Day::Monday,
            "tuesday" => Day::Tuesday,
            "wednesday" => Day::Wednesday,
            "thursday" => Day::Thursday,
            "friday" => Day::Friday,
            "saturday" => Day::Saturday,
            "sunday" => Day::Sunday,
            _ => anyhow::bail!("Unrecognized day {s}"),
        };
        Ok(day)
    }
}

impl From<&Day> for u8 {
    fn from(day: &Day) -> Self {
        day.clone() as u8 + 1
    }
}

impl From<u8> for Day {
    fn from(day: u8) -> Self {
        let week = [
            Day::Monday,
            Day::Tuesday,
            Day::Wednesday,
            Day::Thursday,
            Day::Friday,
            Day::Saturday,
            Day::Sunday,
        ];
        week.get(day as usize - 1).expect("Invalid day").clone()
    }
}

impl Display for Day {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let day = match self {
            Day::Monday => "Monday",
            Day::Tuesday => "Tuesday",
            Day::Wednesday => "Wednesday",
            Day::Thursday => "Thursday",
            Day::Friday => "Friday",
            Day::Saturday => "Saturday",
            Day::Sunday => "Sunday",
        };
        write!(f, "{day}")
    }
}
