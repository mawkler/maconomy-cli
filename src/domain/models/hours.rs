pub(crate) struct Hours(pub(crate) f32);

impl From<f32> for Hours {
    fn from(hours: f32) -> Self {
        Self(hours)
    }
}
