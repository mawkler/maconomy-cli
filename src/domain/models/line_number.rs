use std::str::FromStr;

/// 1-indexed number for line in time sheet
#[derive(Debug, Clone)]
pub(crate) enum LineNumber {
    Number(u8),
    Last,
}

impl FromStr for LineNumber {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let line_number = match s {
            "last" => Self::Last,
            _ => {
                let number = s
                    .parse()
                    .map_err(|err| format!("Failed to parse number: {err}"))?;
                Self::Number(number)
            }
        };

        match line_number {
            Self::Number(0) => {
                Err("Invalid line number. Note that line numbers are 1-based".to_string())
            }
            n => Ok(n),
        }
    }
}
