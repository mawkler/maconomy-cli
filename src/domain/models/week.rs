use anyhow::anyhow;
use chrono::{Datelike, NaiveDate, Weekday};
use std::fmt::Display;

#[derive(Debug, Clone, Default, serde::Serialize)]
pub(crate) struct WeekNumber {
    number: u8,
    year: i32,
}

impl WeekNumber {
    pub(crate) fn new(week: u8, year: i32) -> anyhow::Result<Self> {
        first_day_of_week(week, year).ok_or(anyhow!("Invalid week '{week}'"))?;

        Ok(Self { number: week, year })
    }

    fn new_with_year_fallback(week: u8, year: Option<i32>) -> anyhow::Result<Self> {
        // Fall back to today's year
        let year = year.unwrap_or_else(|| chrono::Utc::now().year());
        WeekNumber::new(week, year)
    }

    pub(crate) fn new_with_fallback(
        week: Option<u8>,
        year: Option<i32>,
    ) -> anyhow::Result<WeekNumber> {
        let week = week.unwrap_or_else(|| {
            // Fall back to today's week
            let this_week = chrono::Local::now().date_naive().iso_week().week();
            this_week
                .try_into()
                .expect("Week numbers are always less than 255")
        });

        WeekNumber::new_with_year_fallback(week, year)
    }

    pub(crate) fn first_day(&self) -> Option<NaiveDate> {
        first_day_of_week(self.number, self.year)
    }
}

impl Display for WeekNumber {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Week {}, year {}", self.number, self.year)
    }
}

fn first_day_of_week(week: u8, year: i32) -> Option<NaiveDate> {
    NaiveDate::from_isoywd_opt(year, week.into(), Weekday::Mon)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn gets_date_of_first_day_of_week() {
        let dates = [
            (2024, 46, "2024-11-11"),
            (2030, 1, "2029-12-31"),
            (2028, 23, "2028-06-05"),
        ];

        for (year, week, expected_date) in dates {
            let week_number = WeekNumber { number: week, year };
            let date = week_number.first_day().unwrap();

            assert_eq!(
                date,
                NaiveDate::parse_from_str(expected_date, "%Y-%m-%d").unwrap()
            )
        }
    }
}
