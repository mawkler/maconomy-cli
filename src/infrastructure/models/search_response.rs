use serde::{Deserialize, Serialize};

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SearchResponse<T> {
    pub(crate) panes: Panes<T>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Panes<T> {
    pub(crate) filter: Filter<T>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Filter<T> {
    pub(crate) meta: Meta,
    pub(crate) records: Vec<Record<T>>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Meta {
    pub(crate) pane_name: String,
    pub(crate) row_count: i64,
    pub(crate) row_offset: i64,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Record<T> {
    pub(crate) data: T,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Tasks {
    pub(crate) taskname: String,
    pub(crate) tasklist: String,
    pub(crate) description: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Jobs {
    pub(crate) jobnumber: String,
}
