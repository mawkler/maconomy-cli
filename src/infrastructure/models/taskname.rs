/// Maconomy tasks have two different names, one called `taskname` (which seems to be shorter) and
/// one called `tasktextvar` which seems to be a longer name for the same task. Our domain uses the
/// longer name (i.e. `tasktextvar`).
///
/// This type represents a `taskname`.
pub(crate) struct ShortTaskName(pub(crate) String);
