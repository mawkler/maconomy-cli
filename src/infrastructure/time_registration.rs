use serde::Deserialize;
use serde::Serialize;

// This file was (semi) auto-generated based on the JSON response from Maconomy. I just improved the struct names.

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimeRegistration {
    pub meta: Meta,
    pub panes: Panes,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Meta {
    pub container_instance_id: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Panes {
    pub card: Card,
    pub table: Table,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Card {
    pub meta: CardMeta,
    pub records: Vec<CardRecord>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CardMeta {
    pub pane_name: String,
    pub row_count: u32,
    pub row_offset: i64,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CardRecord {
    pub data: CardData,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CardData {
    pub employeenumber: String,
    pub periodstartvar: String,
    pub periodendvar: String,
    pub employeenamevar: String,
    pub datevar: String,
    pub weeknumbervar: u8,
    pub fixednumberday1var: f32,
    pub fixednumberday2var: f32,
    pub fixednumberday3var: f32,
    pub fixednumberday4var: f32,
    pub fixednumberday5var: f32,
    pub fixednumberday6var: f32,
    pub fixednumberday7var: f32,
    pub totalnumberday1var: f32,
    pub totalnumberday2var: f32,
    pub totalnumberday3var: f32,
    pub totalnumberday4var: f32,
    pub totalnumberday5var: f32,
    pub totalnumberday6var: f32,
    pub totalnumberday7var: f32,
    pub regulartimeday1var: f32,
    pub regulartimeday2var: f32,
    pub regulartimeday3var: f32,
    pub regulartimeday4var: f32,
    pub regulartimeday5var: f32,
    pub regulartimeday6var: f32,
    pub regulartimeday7var: f32,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Table {
    pub meta: TableMeta,
    pub records: Vec<TableRecord>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableMeta {
    pub pane_name: String,
    pub row_count: u32,
    pub row_offset: i64,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableRecord {
    pub data: TableData,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableData {
    pub jobnumber: String,
    pub numberday1: f32,
    pub numberday2: f32,
    pub numberday3: f32,
    pub numberday4: f32,
    pub numberday5: f32,
    pub numberday6: f32,
    pub numberday7: f32,
    pub entrytext: String,
    pub taskname: String,
    pub instancekey: String,
    pub timeregistrationunit: String,
    pub jobnamevar: String,
    pub tasktextvar: String,
}
