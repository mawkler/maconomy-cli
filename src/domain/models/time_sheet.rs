use super::hours::Hours;

#[derive(Debug, serde::Serialize)]
pub(crate) struct Week {
    pub(crate) monday: Hours,
    pub(crate) tuesday: Hours,
    pub(crate) wednesday: Hours,
    pub(crate) thursday: Hours,
    pub(crate) friday: Hours,
    pub(crate) saturday: Hours,
    pub(crate) sunday: Hours,
}

#[derive(Debug, serde::Serialize)]
pub(crate) struct Line {
    pub(crate) job: String,
    pub(crate) task: String,
    pub(crate) week: Week,
}

impl Line {
    pub(crate) fn new(job: String, task: String, week: Week) -> Self {
        Self { job, task, week }
    }
}

#[derive(Default, serde::Serialize)]
pub(crate) struct TimeSheet {
    pub(crate) lines: Vec<Line>,
}

impl TimeSheet {
    pub(crate) fn new(lines: Vec<Line>) -> Self {
        Self { lines }
    }
}

impl TimeSheet {
    pub(crate) fn find_line_nr(&self, job: &str, task: &str) -> Option<u8> {
        let (row, _) = self
            .lines
            .iter()
            .enumerate()
            .find(|(_, line)| line.job == job && line.task == task)?;

        Some(row as u8)
    }
}
