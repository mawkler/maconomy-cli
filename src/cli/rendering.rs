use crate::infrastructure::models::time_registration;
use owo_colors::OwoColorize;
use std::fmt::Display;
use tabled::settings::{
    object::Rows, style::BorderColor, themes::Colorization, Color, Style, Theme,
};

#[derive(tabled::Tabled, Default)]
pub(crate) struct Row<'a> {
    #[tabled(rename = "Job name")]
    pub(crate) job_name: &'a str,
    #[tabled(rename = "Task name")]
    pub(crate) task_name: &'a str,
    #[tabled(rename = "Mon")]
    #[tabled(display_with = "display_hours")]
    pub(crate) monday: f32,
    #[tabled(rename = "Tue")]
    #[tabled(display_with = "display_hours")]
    pub(crate) tuesday: f32,
    #[tabled(rename = "Wed")]
    #[tabled(display_with = "display_hours")]
    pub(crate) wednesday: f32,
    #[tabled(rename = "Thu")]
    #[tabled(display_with = "display_hours")]
    pub(crate) thursday: f32,
    #[tabled(rename = "Fri")]
    #[tabled(display_with = "display_hours")]
    pub(crate) friday: f32,
    #[tabled(rename = "Sat")]
    #[tabled(display_with = "display_hours")]
    pub(crate) saturday: f32,
    #[tabled(rename = "Sun")]
    #[tabled(display_with = "display_hours")]
    pub(crate) sunday: f32,
}

fn display_hours(hours: &f32) -> impl Display {
    if let 0.0 = hours {
        return "".to_string();
    }

    let whole_hours = hours.trunc() as u32;
    let minutes = ((*hours - whole_hours as f32) * 60.0).floor() as u32;

    format!("{whole_hours}:{minutes:02}")
}

impl<'a> From<&'a time_registration::TableRecord> for Row<'a> {
    fn from(record: &'a time_registration::TableRecord) -> Self {
        Row {
            job_name: &record.data.jobnamevar,
            task_name: &record.data.tasktextvar,
            monday: record.data.numberday1,
            tuesday: record.data.numberday2,
            wednesday: record.data.numberday3,
            thursday: record.data.numberday4,
            friday: record.data.numberday5,
            saturday: record.data.numberday6,
            sunday: record.data.numberday7,
        }
    }
}

fn gray_borders() -> BorderColor {
    let gray = Color::parse(' '.fg_rgb::<85, 85, 85>().to_string()).clone();

    BorderColor::new()
        .top(gray.clone())
        .left(gray.clone())
        .bottom(gray.clone())
        .right(gray.clone())
        .corner_bottom_right(gray.clone())
        .corner_bottom_left(gray.clone())
        .corner_top_left(gray.clone())
        .corner_top_right(gray.clone())
}

impl Display for time_registration::TimeRegistration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let rows = self.panes.table.records.iter().map(Row::from);

        let mut theme = Theme::from_style(Style::modern_rounded());
        theme.remove_vertical_lines();

        let mut table = tabled::Table::new(rows);
        let table = table
            .with(theme)
            .with(BorderColor::new().top(Color::FG_GREEN))
            .with(Colorization::exact(
                [tabled::settings::Color::BOLD],
                Rows::first(),
            ))
            .with(gray_borders());

        write!(f, "{table}")
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn display_hours_test() {
        let hours = [1.5, 0.25, 12.75, 0.0, 23.999, 10.1];
        let expected = ["1:30", "0:15", "12:45", "", "23:59", "10:06"];

        for (i, &hours) in hours.iter().enumerate() {
            let result = display_hours(&hours).to_string();
            assert_eq!(result, expected[i], "Failed at index {}", i);
        }
    }
}
