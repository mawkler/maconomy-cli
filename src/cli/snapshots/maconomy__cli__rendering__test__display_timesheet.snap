---
source: src/cli/rendering.rs
expression: time_sheet.to_string()
snapshot_kind: text
---
[38;2;85;85;85mâ•­â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m
[38;2;85;85;85mâ”‚[39m[1m [22m[1mJob name[22m[1m        [22m[1m [22m[1m [22m[1mTask name[22m[1m        [22m[1m [22m[1m [22m[1mMon[22m[1m [22m[1m [22m[1m [22m[1mTue[22m[1m [22m[1m [22m[1m [22m[1mWed[22m[1m [22m[1m [22m[1m [22m[1mThu[22m[1m [22m[1m [22m[1m [22m[1mFri[22m[1m [22m[1m [22m[1m [22m[1mSat[22m[1m [22m[1m [22m[1m [22m[1mSun[22m[1m [22m[1m [22m[38;2;85;85;85mâ”‚[39m
[38;2;85;85;85mâ”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤[39m
[38;2;85;85;85mâ”‚[39m Job number one    Task number one    8:00  8:00                               [38;2;85;85;85mâ”‚[39m
[38;2;85;85;85mâ”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤[39m
[38;2;85;85;85mâ”‚[39m job number two    task number two                8:00  8:00  1:00  1:00       [38;2;85;85;85mâ”‚[39m
[38;2;85;85;85mâ”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤[39m
[38;2;85;85;85mâ”‚[39m job number three  task number three                          7:00  7:00  8:00 [38;2;85;85;85mâ”‚[39m
[38;2;85;85;85mâ”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤[39m
[38;2;85;85;85mâ”‚[39m[38;2;85;85;85m [39m[38;2;85;85;85mWeek 47[39m[38;2;85;85;85m                                                                      [39m[38;2;85;85;85m [39m[38;2;85;85;85mâ”‚[39m
[38;2;85;85;85mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m
